### PR TITLE
Use sharded tokens in MSC4354: Sticky Events sync and sliding sync, preventing events appearing in multiple sync responses.

### LINKED ISSUE #19661: MSC4354 Sticky Events sync and sliding sync should use a sharded token
Experimental feature tracking issue: #19409

Spawning from https://github.com/element-hq/synapse/pull/19365#discussion_r3017099461

---

When returning sticky events over sync and sliding sync, we currently use a single int as a stream token, but should change to using a sharded stream token.

As a very basic recap of https://element-hq.github.io/synapse/latest/development/synapse_architecture/streams.html#current-stream-id: a stream reader that uses a single int on a multi-writer stream must wait for all writers to finish persisting their work before advancing up to that position.
This means that the reader can be delayed in reading new facts in the stream.

Since it's event persistence that writes to both streams, this will probably cause a slightly undesirable situation where the event first appears in the timeline before appearing in the sticky stream (from the point of view of a stream reader serving sync).
As a result, users may see the sticky event sent down two (e.g.) consecutive sync responses, once in the timeline and once in the sticky section. This defeats the 'deduplication' logic intended in MSC4354, so we ideally want these two streams to be perfectly synchronised.

In other words, we need to use sharded tokens on the sticky events stream as well.

### PR DESCRIPTION
**Pull request by reivilibre**
Fixes: #19661
Part of: MSC4354

This pull request is commit-by-commit review friendly. 

---

This PR first introduces some helpers for querying a sharded stream:

- SQL builder for 'give me rows between two sharded tokens'
- a function to return the 'new' from_token after a limited read

It's expected that these helpers will be useful in other places, such as https://github.com/element-hq/synapse/pull/20108#discussion_r3924032994
Some tests are included.

Then, using those helpers, I make the sticky events portions of both
oldschool and sliding sync use sharded tokens.

This prevents a 'torn sync' problem where, because the events and sticky events
streams are both sharded with the same writers:

1. persisting a sticky event leads to an event row and a sticky_events row being created for event `e` (on writer `W`).
2. Suppose another writer is still persisting, so the position of the stream for writer W advances, but the global ['linear'](https://element-hq.github.io/synapse/latest/development/synapse_architecture/streams.html?highlight=%E2%80%9Clinear%E2%80%9D#current-stream-id) (I have been calling this 'baseline') position does not advance yet.
3. sync is sharding-aware for the regular events stream, so `e` appears as a real-time timeline event, thus coming down sync
4. since sticky events were not sharding-aware, `e` is not detected as a sticky event in this sync response. The sticky event position doesn't advance in the sync token, either.
5. (assume the position advances because other writers have finished writing and so the 'linear' position advances for the sticky events stream)
6. the client then syncs again, using the new sync token. This time, `e` appears in the sticky event section.

If we make sticky events sharding-aware in sync, then `e` would appear for consideration in both the timeline section **and** the sticky event section in the first sync.
It then gets deduplicated out of the sticky section since it is appearing in the timeline section in the same sync.

Original commit schedule, with full messages:

<ol>
<li>

Add utilities for sharded stream tokens 

</li>
<li>

Support sharded tokens for sticky events in sync 

</li>
<li>

Update storage test to follow rename 

</li>
<li>

Add sharding tests for sticky events \
Conceptual cousin of `test_sharded_receipts` etc

Main interesting case we want to prevent is the 'torn sync' problem, so
that's what I've focussed on for the test.
It is sufficient to demonstrate that the stream is sharded.

</li>
<li>

Comment on MultiWriterStreamToken.is_stream_position_in_range 

</li>
<li>

Add storage test against sticky event token rewind 

</li>
</ol>

### CODE REVIEW

File: synapse/replication/tcp/client.py [hunk 1]
```diff
@@ -286,7 +286,7 @@ async def on_rdata(
         elif stream_name == StickyEventsStream.NAME:
             self.notifier.on_new_event(
                 StreamKeyType.STICKY_EVENTS,
-                token,
+                self.store.get_sticky_events_stream_token(),
```
Comment (reivilibre) [inline synapse/replication/tcp/client.py:289]: I believe the old form, of passing in the RDATA's token, instead of the stream's actual current position was a mistake.
See https://github.com/element-hq/synapse/issues/19969

File: synapse/storage/util/id_generators.py [hunk 2]
```diff
@@ -36,6 +36,7 @@
 )
 
 import attr
+from immutabledict import immutabledict
```
Comment (sandhose) [inline synapse/storage/util/id_generators.py:39]: I was surprised to see `immutabledict` instead of say `frozendict` which is builtin, but it looks like we're already using that all over the place. So that's for a future refactor ⏩

File: synapse/storage/util/id_generators.py [hunk 3]
```diff
@@ -982,3 +984,180 @@ async def __aexit__(
...
+
+    Bounds: from_token_exclusive < ... <= to_token_inclusive
+
+    Arguments:
+        stream_id_column: the name of the `stream_id` column
+            Should be prefixed with the table name or alias
+            when used in a multi-table SELECT statement.
+        instance_name_column: the name of the `instance_name` column.
+            Should be prefixed with the table name or alias
+            when used in a multi-table SELECT statement.
```
Comment (sandhose) [inline synapse/storage/util/id_generators.py:1008]: I believe we almost always use `instance_name` for this, so alternatively we *could* ask for the `table_name` instead, and automatically do `{table_name}.instance_name` / `{table_name}.{stream_id_column}`? But maybe that's too magic and doing too many assumptions

File: synapse/storage/util/id_generators.py [hunk 4]
```diff
@@ -982,3 +984,180 @@ async def __aexit__(
...
+    for instance_name, pos in to_token_inclusive.instance_map.items():
+        upper_or_clauses.append(
+            f"({instance_name_column} = ? AND {stream_id_column} <= ?)"
+        )
+        upper_or_values.extend((instance_name, pos))
+
+    clauses.append("(\n\t\t" + "\n\t\tOR ".join(upper_or_clauses) + "\n\t)")
+    values.extend(upper_or_values)
+
+    return "(\n\t" + "\n\tAND ".join(clauses) + "\n)", values
```
Comment (sandhose) [inline synapse/storage/util/id_generators.py:1090]: Not a fan of this query building with newlines and tabs but oh well :D

File: synapse/storage/util/id_generators.py [hunk 5]
```diff
@@ -982,3 +984,180 @@ async def __aexit__(
...
+        to_token_inclusive: the inclusive upper bound the rows were read with.
+        last_read_stream_id: the `stream_id` of the last row that was read.
+
+    Returns:
+        A token that could be used as the `from_token_exclusive` for a subsequent
+        read, in order to continue reading rows in range.
+
+        The token represents the maximum 'read up to' position.
+        Rows with positions strictly above the token have not yet been read.
+    """
```
Comment (sandhose) [inline synapse/storage/util/id_generators.py:1118]: I think I understand what this helper does, but I think there's another assumption that isn't spelled out here: what about when you do a partial read, but stop in the 'middle' of a stream_id? As in, in a multi-writer scenario, it's possible that two writers write with the same `stream_id` but different `instance_name`, right? So what if the partial read decides to stop in the "middle" of the same `stream_id`, if that makes sense?

File: synapse/storage/util/id_generators.py (hunk 3 above)
Comment (sandhose) [inline synapse/storage/util/id_generators.py:1008]: I also believe this method only works if `instance_name` is `NOT NULL` which isn't always the case: we have a bunch of streams where `NULL` means "written before we were multi-stream", which is the case for most streams. So either fix that so that it works for rows that are `NULL`, or put a prerequisite in the docstring saying the `instance_name` must not contain `NULL` values

File: synapse/storage/databases/main/sticky_events.py [hunk 6]
```diff
@@ -141,65 +146,84 @@ async def get_sticky_events_in_rooms(
...
             now: The current time in unix millis, used for skipping expired events.
             limit: Max sticky events to return, or None to apply no limit.
         Returns:
-            to_id, dict[room_id, list[event_ids]]
+            The stream position that has been read up to (which may be behind
+            `to_token` if `limit` was hit), and a dict[room_id, list[event_ids]].
         """
+        if limit == 0:
+            # No rows to return, so don't advance.
+            return from_token, {}
```
Comment (sandhose) [inline synapse/storage/databases/main/sticky_events.py:172]: I think you're fixing a real potential bug here, right? As in, this is something the client can reach that with basically `StickyEventsExtension.limit = 0`, and before we would just advance the token for the client, because `get_sticky_evnets_in_rooms` would return 0 rows, therefore we would just skip to `to_id`.

Even if it's an edge case, do we have a test for this particular fix?

File: synapse/storage/databases/main/sticky_events.py [hunk 7]
```diff
@@ -141,65 +146,84 @@ async def get_sticky_events_in_rooms(
...
-        # Get stream_id of the last row, which is the highest
-        new_to_id, _, _ = sticky_events_rows[-1]
+        if limit is not None and len(sticky_events_rows) == limit:
+            # We hit the limit, so advance the `to_token` partially
+            last_stream_id, _, _ = sticky_events_rows[-1]
+            new_to_token = advance_multiwriter_sharded_token_after_partial_read(
+                from_token_exclusive=from_token,
+                to_token_inclusive=to_token,
+                last_read_stream_id=last_stream_id,
+            )
```
Comment (sandhose) [inline synapse/storage/databases/main/sticky_events.py:191]: I believe here we could actually hit the bug I was talking about: what if the limit gets us in the middle of a `stream_id`? I think we should read all the last rows that were applied where `stream_id === last_stream_id` and take those into account to compute an accurate `new_to_token`?

File: synapse/storage/databases/main/sticky_events.py [hunk 8]
```diff
@@ -141,65 +146,84 @@ async def get_sticky_events_in_rooms(
...
         *,
-        from_id: int,
-        to_id: int,
+        from_token: MultiWriterStreamToken,
+        to_token: MultiWriterStreamToken,
         now: int,
         limit: int | None,
     ) -> list[tuple[int, str, str]]:
-        if len(room_ids) == 0:
+        if len(room_ids) == 0 or limit == 0 or to_token.is_before_or_eq(from_token):
```
Comment (sandhose) [inline synapse/storage/databases/main/sticky_events.py:216]: I believe `limit == 0` is dead here, it's checked by the caller?

`test_get_sticky_events_in_rooms_token_does_not_go_backwards` explains a little bit why we're checking that `to_token <= from_token`, but it wouldn't hurt commenting about this here

Something like "A client can hand us a token that is ahead of our own stream positions, our own 'now token'"?

File: synapse/storage/util/id_generators.py (hunk 5 above)
Comment (sandhose) [reply synapse/storage/util/id_generators.py:1118]: Re-reading this, the stream allocation is always done through the Postgres sequence, so it's unique; but then, maybe we should state in the docstring or something that this works on reads where the stream_id *is* unique. In some places I believe we assign the same stream_id for multiple rows?

File: synapse/storage/util/id_generators.py (hunk 2 above)
Comment (reivilibre) [reply synapse/storage/util/id_generators.py:39]: `frozendict` is only in Py 3.15 it seems https://peps.python.org/pep-0814/

File: synapse/storage/util/id_generators.py (hunk 3 above)
Comment (reivilibre) [reply synapse/storage/util/id_generators.py:1045]: I have a very slight preference for passing it in explicitly so you can see it being used even if you don't know what the function does

File: synapse/storage/util/id_generators.py (hunk 5 above)
Comment (reivilibre) [reply synapse/storage/util/id_generators.py:1165]: > it's possible that two writers write with the same stream_id but different instance_name, right? So what if the partial read decides to stop in the "middle" of the same stream_id, if that makes sense?

No, this isn't possible

When you have multiple rows with the same `stream_id` they are all considered part of the same 'fact' and should be persisted together (so by the same worker) and should also basically always be processed/read together https://element-hq.github.io/synapse/latest/development/synapse_architecture/streams.html?highlight=same%20stream_id#definition

But probably is for the best to point this out on the function docstring, yeah ae17138ec7fe06b1f766d9b6f151ad7780a6f7bf

File: synapse/storage/util/id_generators.py (hunk 2 above)
Comment (sandhose) [reply synapse/storage/util/id_generators.py:39]: …why did I even think frozendict was already a thing 🥲

File: synapse/storage/util/id_generators.py (hunk 4 above)
Comment (reivilibre) [reply synapse/storage/util/id_generators.py]: Yeah mrm, I was torn between this or having a helper to assertEqual whilst tolerating the whitespace differences like in 78bd2ad9bdb015b199703b66d10373f7ddfbe7ea

File: synapse/storage/databases/main/sticky_events.py (hunk 7 above)
Comment (reivilibre) [reply synapse/storage/databases/main/sticky_events.py:200]: For sticky events, `stream_id` is only used once per fact, so can't happen here

File: synapse/storage/util/id_generators.py (hunk 3 above)
Comment (reivilibre) [reply synapse/storage/util/id_generators.py:1045]: Very good spot. I believe 38ddf6a47d78ee8f60aba44159af90932d0794d9 tested by a47b8ffeb1b2088c6bcb222f342fd13b034964fd sorts that one out

File: synapse/storage/databases/main/sticky_events.py (hunk 6 above)
Comment (reivilibre) [reply synapse/storage/databases/main/sticky_events.py:181]: Because no sticky events get returned, the `next_batch` doesn't get returned either, so we were saved by this technicality
https://github.com/element-hq/synapse/blob/e8987e801ecdbdc8ccfa6819db40b2baf25265f8/synapse/types/handlers/sliding_sync.py#L443

But added a test in 3a3e22d9834c43bae2fc8ec21e987cc41849bc9c

File: synapse/storage/util/id_generators.py (hunk 4 above)
Comment (sandhose) [reply synapse/storage/util/id_generators.py]: Didn't realize this was for asserting in tests, it's a good reason to have this well formatted 👍

File: tests/storage/test_id_generators.py [hunk 9]
```diff
@@ -915,3 +921,328 @@ def test_load_existing_stream(self) -> None:
...
+
+    These don't need a homeserver: they only exercise the SQL that
+    `make_multiwriter_sharded_token_bounds_sql` builds, against a throwaway
+    SQLite table.
+
+    NOTE: `ShardedTokenHelpersDatabaseNullTestCase` inherits all these tests
+        with a separate set of tables.
+    """
+
+    ROWS = [
```
Comment (sandhose) [inline tests/storage/test_id_generators.py]: this is inferred as `tuple[int, str]` which then fails when you insert `(4, None)` later. That's why mypy is currently failing
```suggestion
    ROWS: tuple[int, str | None] = [
```

File: tests/replication/test_sharded_sticky_events.py [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+    def _create_room(self, room_id: str) -> None:
+        """
+        Create a room with the given room ID, so that we control which event
+        persister ends up owning it.
+        """
+        with patch(
+            "synapse.handlers.room.RoomCreationHandler._generate_room_id"
+        ) as mock:
+            mock.side_effect = lambda: room_id
+            self.helper.create_room_as(self.user_id, tok=self.tok)
```
Comment (sandhose) [inline tests/replication/test_sharded_sticky_events.py]: since the develop merge and #20130 got in there, I think this isn't the way of doing things anymore 🤔 

https://github.com/element-hq/synapse/pull/20130/changes#diff-e5101d5a05d517713015ae9ae42938bcf517b570c06f3ec5d2c94aabc64173a4L66

File: synapse/storage/util/id_generators.py [hunk 11]
```diff
@@ -1019,3 +1021,190 @@ async def __aexit__(
...
+        values.extend((instance_name, pos))
+
+    # Now we need to restrict the upper end of the envelope to only include those rows where either:
+    # - the row is in a region where all writers' rows are visible; or
+    upper_or_clauses = [f"{stream_id_column} <= ?"]
+    upper_or_values: list[int | str] = [to_token_inclusive.stream]
+    # - the row's writer is explicitly visible ahead of the baseline (minimum) position
+    for instance_name, pos in to_token_inclusive.instance_map.items():
+        upper_or_clauses.append(
+            f"({instance_name_column} IS NOT DISTINCT FROM ? AND {stream_id_column} <= ?)"
```
Comment (sandhose) [inline synapse/storage/util/id_generators.py]: You are going to *love* this: `IS NOT DISTINCT FROM` was introduced in [SQLite 3.39.0](https://sqlite.org/releaselog/3_39_0.html). We [check at runtime SQLite >= 3.37.2](https://github.com/element-hq/synapse/blob/8f6f4e03c18248a9f659464a7b09640bc4a8f906/synapse/storage/engines/sqlite.py#L79). We did say ["SQLite 3.40.0 or later"](https://github.com/element-hq/synapse/blob/develop/docs/upgrade.md?plain=1#L296) about [a year ago](https://github.com/element-hq/synapse/releases/tag/v1.143.0) but our "trial-olddeps" job [runs on Ubuntu 22.04](https://github.com/element-hq/synapse/blob/8f6f4e03c18248a9f659464a7b09640bc4a8f906/.github/workflows/tests.yml#L471) which means [SQLite 3.37.2](https://launchpad.net/ubuntu/jammy/+source/sqlite3)

Not sure what to do with all that information :D

Maybe raise the runtime check? Or change the query

File: tests/rest/client/sliding_sync/test_extension_sticky_events.py [hunk 12]
```diff
@@ -594,6 +594,68 @@ def test_sticky_event_pagination(self) -> None:
             response_body, {room_id: sticky_event_ids[2:4]}
         )
 
+    def test_sticky_event_limit_zero_does_not_advance(self) -> None:
+        """
+        Tests that sending a limit of 0 in the Sticky Events request extension doesn't
+        advance the sticky events stream token to a future position.
```
Comment (sandhose) [inline tests/rest/client/sliding_sync/test_extension_sticky_events.py]: I feel like this would be more accurate? As we don't check the next_batch token

```suggestion
        Tests that sending a limit of 0 in the Sticky Events request extension returns no
        sticky events and omits the extension from the response, rather than erroring.
```

File: tests/storage/test_id_generators.py (hunk 9 above)
Comment (reivilibre) [reply tests/storage/test_id_generators.py]: derp Did I really forget to check mypy

File: synapse/storage/util/id_generators.py (hunk 11 above)
Comment (reivilibre) [reply synapse/storage/util/id_generators.py]: We already have some fun with SQLite versions (see https://github.com/element-hq/synapse/issues/19452)
but that's fine, not too hard to use the old spelling `IS` instead of `IS NOT DISTINCT FROM` on SQLite for now...

File: tests/replication/test_sharded_sticky_events.py (hunk 10 above)
Comment (reivilibre) [reply tests/replication/test_sharded_sticky_events.py]: ty for the hint!
